beacon-chain/blockchain/receive_execution_payload_envelope.go

```diff
@@ -0,0 +1,289 @@
+	blockHeader *ethpb.BeaconBlockHeader,
+	envelope interfaces.ROExecutionPayloadEnvelope,
+) error {
+	if blockHeader.StateRoot == nil || [32]byte(blockHeader.StateRoot) == [32]byte{} {
+		prevStateRoot, err := preState.HashTreeRoot(ctx)
+		if err != nil {
+			return errors.Wrap(err, "could not compute previous state root")
+		}
+		blockHeader.StateRoot = prevStateRoot[:]
+		if err := preState.SetLatestBlockHeader(blockHeader); err != nil {
+			return errors.Wrap(err, "could not set latest block header")
+		}
+	}
+	blockHeaderRoot, err := blockHeader.HashTreeRoot()
+	if err != nil {
+		return err
+	}
+	beaconBlockRoot, err := envelope.BeaconBlockRoot()
+	if err != nil {
+		return err
+	}
+	if blockHeaderRoot != beaconBlockRoot {
+		return errors.New("beacon block root does not match previous header")
+	}
+	return nil
+}
+
+func validateAgainstCommittedBid(
+	committedHeader *enginev1.ExecutionPayloadHeaderEPBS,
+	envelope interfaces.ROExecutionPayloadEnvelope,
+) error {
+	builderIndex, err := envelope.BuilderIndex()
+	if err != nil {
+		return err
+	}
+	if committedHeader.BuilderIndex != builderIndex {
+		return errors.New("builder index does not match committed header")
+	}
+	kzgRoot, err := envelope.BlobKzgCommitmentsRoot()
+	if err != nil {
+		return err
+	}
+	if [32]byte(committedHeader.BlobKzgCommitmentsRoot) != kzgRoot {
+		return errors.New("blob KZG commitments root does not match committed header")
+	}
+	return nil
+}
+
+func processPayloadStateTransition(
+	ctx context.Context,
+	preState state.BeaconState,
+	envelope interfaces.ROExecutionPayloadEnvelope,
+) error {
+	payload, err := envelope.Execution()
+	if err != nil {
+		return err
+	}
+	exe, ok := payload.(interfaces.ExecutionDataElectra)
+	if !ok {
+		return errors.New("could not cast execution data to electra execution data")
+	}
+	preState, err = electra.ProcessDepositRequests(ctx, preState, exe.DepositRequests())
+	if err != nil {
+		return errors.Wrap(err, "could not process deposit receipts")
+	}
+	preState, err = electra.ProcessWithdrawalRequests(ctx, preState, exe.WithdrawalRequests())
+	if err != nil {
+		return errors.Wrap(err, "could not process execution layer withdrawal requests")
+	}
+	if err := electra.ProcessConsolidationRequests(ctx, preState, exe.ConsolidationRequests()); err != nil {
+		return errors.Wrap(err, "could not process consolidation requests")
+	}
+	if err := preState.SetLatestBlockHash(payload.BlockHash()); err != nil {
+		return err
+	}
+	return preState.SetLatestFullSlot(preState.Slot())
+}
+
+func checkPostStateRoot(
+	ctx context.Context,
+	preState state.BeaconState,
+	envelope interfaces.ROExecutionPayloadEnvelope,
+) error {
+	stateRoot, err := preState.HashTreeRoot(ctx)
+	if err != nil {
+		return err
+	}
+	envelopeStateRoot, err := envelope.StateRoot()
+	if err != nil {
+		return err
+	}
+	if stateRoot != envelopeStateRoot {
+		return errors.New("state root mismatch")
+	}
+	return nil
+}
+
+func validatePayloadStateTransition(
```

Comment:
My understanding is this function implements `process_execution_payload` in the consensus spec. If yes, then this is not the right file or folder for this function. Traditionally, we have beacon chain functions inside the `beacon-chain/core` package. We will have spec test vectors for this function, so it will need to be exported as well.

Reply:
Yeah this is right, I was a little afraid of adding yet another file in that package, but you're right, I'll move it out of here.